orderbook/src/solvable_orders.rs

```diff
@@ -0,0 +1,379 @@
+    orderbook::filter_unsupported_tokens,
+};
+use anyhow::Result;
+use model::order::Order;
+use primitive_types::U256;
+use shared::{
+    bad_token::BadTokenDetecting, current_block::CurrentBlockStream, time::now_in_epoch_seconds,
+};
+use std::{
+    collections::{HashMap, HashSet},
+    iter::FromIterator,
+    sync::{Arc, Mutex, Weak},
+    time::{Duration, Instant},
+};
+use tokio::sync::Notify;
+
+/// Keeps track and updates the set of currently solvable orders.
+/// For this we also need to keep track of user sell token balances for open orders so this is
+/// retrievable as well.
+/// The cache is updated in the background whenever a new block appears or when the cache is
+/// explicitly notified that it should update for example because a new order got added to the order
+/// book.
+pub struct SolvableOrdersCache {
+    min_order_validity_period: Duration,
+    database: Arc<dyn OrderStoring>,
+    balance_fetcher: Arc<dyn BalanceFetching>,
+    bad_token_detector: Arc<dyn BadTokenDetecting>,
+    notify: Notify,
+    cache: Mutex<Inner>,
+}
+
+type Balances = HashMap<Query, U256>;
+
+struct Inner {
+    orders: Vec<Order>,
+    balances: Balances,
+    block: u64,
+    update_time: Instant,
+}
+
+impl SolvableOrdersCache {
+    pub fn new(
+        min_order_validity_period: Duration,
+        database: Arc<dyn OrderStoring>,
+        balance_fetcher: Arc<dyn BalanceFetching>,
+        bad_token_detector: Arc<dyn BadTokenDetecting>,
+        current_block: CurrentBlockStream,
+    ) -> Arc<Self> {
+        let self_ = Arc::new(Self {
+            min_order_validity_period,
+            database,
+            balance_fetcher,
+            bad_token_detector,
+            notify: Default::default(),
+            cache: Mutex::new(Inner {
+                orders: Default::default(),
+                balances: Default::default(),
+                block: 0,
+                update_time: Instant::now(),
+            }),
+        });
+        tokio::task::spawn(update_task(Arc::downgrade(&self_), current_block));
+        self_
+    }
+
+    pub fn cached_balance(&self, key: &Query) -> Option<U256> {
+        let inner = self.cache.lock().unwrap();
+        inner.balances.get(key).copied()
+    }
+
+    /// Orders and timestamp at which last update happened.
+    pub fn cached_solvable_orders(&self) -> (Vec<Order>, Instant) {
+        let inner = self.cache.lock().unwrap();
+        (inner.orders.clone(), inner.update_time)
+    }
+
+    /// The cache will update the solvable orders and missing balances as soon as possible.
+    pub fn request_update(&self) {
+        self.notify.notify_one();
+    }
+
+    /// Manually update solvable orders. Usually called by the background updating task.
+    pub async fn update(&self, block: u64) -> Result<()> {
+        let min_valid_to = now_in_epoch_seconds() + self.min_order_validity_period.as_secs() as u32;
+        let orders = self.database.solvable_orders(min_valid_to).await?;
```

Comment:
One other approach I could think of, instead of querying solvable orders, is to use the order `valid_to` for ignoring orders to query balances for.

This way, we could have kept things a little closer to the old model where we:
- "registered" an order for balance updates on creation (registering all solvable orders at startup).
- Only updated orders that are still valid on a new block
- On reorgs (i.e. if the time goes isn't monotonically increasing), we could just re-register all solvable orders.

Reply:
I had the idea too that we could manually add newly created orders instead of fetching from the database but I like this approach because it is stateless. So even if for an unexpected reason the set of solvable orders changes in the database we will notice.